core/src/main/scala/org/apache/spark/ContextCleaner.scala

```diff
@@ -235,8 +235,10 @@ private[spark] class ContextCleaner(
     try {
       if (mapOutputTrackerMaster.containsShuffle(shuffleId)) {
         logDebug("Cleaning shuffle " + shuffleId)
-        mapOutputTrackerMaster.unregisterShuffle(shuffleId)
+        // Shuffle must be removed before it's unregistered from the output tracker
+        // to find blocks served by the shuffle service on deallocated executors
         shuffleDriverComponents.removeShuffle(shuffleId, blocking)
+        mapOutputTrackerMaster.unregisterShuffle(shuffleId)
```

Comment:
Is this new pattern safe in all cases? Any change of `FileNotFound`?

Reply:
It hasn't failed any tests or caused problems in heavy usage for the past couple months. Not sure how else to know

Reply:
@Kimahriman .
- Logically, this is counter intuitive, isn't it? New code is weaker than before.
- I guess you are saying that your experience on your environment and it's usually not Apache Spark master or 3.3. Which forked version are you using now with this patch?

Reply:
Looking at `unregisterShuffle` I don't see anything that would depend on the ShuffleDriverComponents

Reply:
so I agree that this would be better the other way from a logical viewpoint, it makes more sense to unregister before you remove in case it tried to use those shuffle locations and mid removal some were actually there and some were not.  But in the context here normally it would get cleaned if nothing is referencing it anymore, so nothing could try to use it so it shouldn't matter the order.  Now there is also the RDD.cleanShuffleDependencies  though as well.  The problem here I think is he needs map output tracker to figure out what removeShuffle removes.  I haven't looked in detail but it might be better if unregistershuffle could return something that could then be passed to remove Shuffle for instance.

Reply:
Yeah anything trying to use either of these things after cleanup is called sounds like a bug that should be fixed. I'm not sure what the expected behavior of `cleanShuffleDependencies` is while something is actively trying to use the shuffle

Reply:
Also the `removeShuffle` call goes through an RPC call from `BlockManagerMaster` to `BlockManagerMasterEndpoint` so I'm not sure how all that info could be passed

Reply:
Yeah, I'm not sure it really matters, other then possibly the way it fails, either way if someone tries to use the shuffle mid way through the cleanup, things are going to fail.  I would think this is fine.